staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -130,32 +157,35 @@ func (q *delayingType) waitingLoop() {
 	// Make a placeholder channel to use when there are no items in our list
 	never := make(<-chan time.Time)
 
+	waitingForQueue := &waitForPriorityQueue{}
+	heap.Init(waitingForQueue)
+
+	waitingEntryByData := map[t]*waitFor{}
+
 	for {
 		if q.Interface.ShuttingDown() {
-			// discard waiting entries
-			q.waitingForAdd = nil
-			q.waitingTimeByEntry = nil
 			return
 		}
 
 		now := q.clock.Now()
 
 		// Add ready entries
-		readyEntries := 0
-		for _, entry := range q.waitingForAdd {
+		for waitingForQueue.Len() > 0 {
+			entry := waitingForQueue.Peek().(*waitFor)
```

Comment:
@deads2k Gotcha, I agree this initially looks confusing but I think it's correct. I'll try to explain it as best I can here, but let me know if you think I can clarify or make the code more clear.

* The [container/heap docs](https://godoc.org/container/heap) state that "The minimum element in the tree is the root, at index 0."
* But the [container/heap#Interface docs] state that the `Pop` method should "remove and return element Len() - 1." and that the `Push` method should "add x as element Len()"

Basically, the 0th element of the heap is the least element (in this case, the element with the timestamp that will occur next in time). Therefore that's the element we want to `Peek`. In the `Pop` case, though, the `container/heap` [will have moved the minimum element to be temporarily at `Len() - 1`](https://golang.org/src/container/heap/heap.go#L61), so we pop off from that end. In the `Push` case, we add the new element to the 'end' of the heap and the element is percolated up to its correct position by `container/heap`.

I've [created a simple integer-based heap on the playground](https://play.golang.org/p/vK5fmN_vfD) that might demonstrate this more clearly.